### PR TITLE
fix shell checks errors in cluster/common.sh

### PR DESCRIPTION
**Pull request by beautytiger**
**What type of PR is this?**
/kind bug
/kind cleanup

**What this PR does / why we need it**:
cheanup shellcheck failings in cluster/common.sh

**Which issue(s) this PR fixes**:

Ref #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

/cc @BenTheElder
/assign @BenTheElder

### CODE REVIEW

File: cluster/common.sh [hunk 1]
```diff
@@ -456,7 +468,7 @@ EOF
       exit 2
   esac
 
-  popd
+  popd "$@"
```
Comment (BenTheElder) [inline cluster/common.sh]: this change seems suspect

File: cluster/common.sh [hunk 2]
```diff
@@ -358,6 +367,9 @@ function generate-etcd-cert() {
   local GEN_ETCD_CA_CERT=${GEN_ETCD_CA_CERT:-}
   local GEN_ETCD_CA_KEY=${GEN_ETCD_CA_KEY:-}
 
+  local ca_cert=${ca_cert:-}
+  local ca_key=${ca_key:-}
```
Comment (BenTheElder) [inline cluster/common.sh]: why this change?

File: cluster/common.sh [hunk 3]
```diff
@@ -322,15 +328,18 @@ function find-tar() {
...
   if [[ "${NUM_WINDOWS_NODES}" -gt "0" ]]; then
+    # shellcheck disable=SC2034
     NODE_BINARY_TAR=$(find-tar kubernetes-node-windows-amd64.tar.gz)
   fi
 
   # This tarball is used by GCI, Ubuntu Trusty, and Container Linux.
   KUBE_MANIFESTS_TAR=
   if [[ "${MASTER_OS_DISTRIBUTION:-}" == "trusty" || "${MASTER_OS_DISTRIBUTION:-}" == "gci" || "${MASTER_OS_DISTRIBUTION:-}" == "ubuntu" ]] || \
      [[ "${NODE_OS_DISTRIBUTION:-}" == "trusty" || "${NODE_OS_DISTRIBUTION:-}" == "gci" || "${NODE_OS_DISTRIBUTION:-}" == "ubuntu" || "${NODE_OS_DISTRIBUTION:-}" == "custom" ]] ; then
+    # shellcheck disable=SC2034
```
Comment (BenTheElder) [inline cluster/common.sh]: why?

File: cluster/common.sh [hunk 4]
```diff
@@ -322,15 +328,18 @@ function find-tar() {
 #   SERVER_BINARY_TAR
 #   KUBE_MANIFESTS_TAR
 function find-release-tars() {
+  # shellcheck disable=SC2034
   SERVER_BINARY_TAR=$(find-tar kubernetes-server-linux-amd64.tar.gz)
   if [[ "${NUM_WINDOWS_NODES}" -gt "0" ]]; then
+    # shellcheck disable=SC2034
```
Comment (BenTheElder) [inline cluster/common.sh]: why?

File: cluster/common.sh [hunk 5]
```diff
@@ -322,15 +328,18 @@ function find-tar() {
 #   SERVER_BINARY_TAR
 #   KUBE_MANIFESTS_TAR
 function find-release-tars() {
+  # shellcheck disable=SC2034
```
Comment (BenTheElder) [inline cluster/common.sh]: why?

File: cluster/common.sh [hunk 6]
```diff
@@ -286,6 +291,7 @@ function set_binary_version() {
   else
     KUBE_VERSION=${1}
   fi
+  export KUBE_VERSION
```
Comment (BenTheElder) [inline cluster/common.sh]: do we actually want this exported? do other processes read it?
update the doc comment if so to clarify that it is exported

File: cluster/common.sh [hunk 7]
```diff
@@ -211,7 +214,7 @@ function get-kubeconfig-user-basicauth() {
 #   KUBE_USER
 #   KUBE_PASSWORD
 function gen-kube-basicauth() {
-    KUBE_USER=admin
+    KUBE_USER="admin"
```
Comment (BenTheElder) [inline cluster/common.sh]: we should prefer single quotes for literals

File: cluster/common.sh [hunk 8]
```diff
@@ -29,15 +29,15 @@ source "${KUBE_ROOT}/hack/lib/util.sh"
...
+export KUBE_RELEASE_VERSION_DASHED_REGEX="v(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-(0|[1-9][0-9]*)(-([a-zA-Z0-9]+)-(0|[1-9][0-9]*))?"
 
 # KUBE_CI_VERSION_REGEX matches things like "v1.2.3-alpha.4.56+abcdefg" This
 #
 # NOTE This must match the version_regex in build/common.sh
 # kube::release::parse_and_validate_ci_version()
-KUBE_CI_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)-([a-zA-Z0-9]+)\\.(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)\\+[-0-9a-z]*)?$"
-KUBE_CI_VERSION_DASHED_REGEX="^v(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-([a-zA-Z0-9]+)-(0|[1-9][0-9]*)(-(0|[1-9][0-9]*)\\+[-0-9a-z]*)?"
+export KUBE_CI_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)-([a-zA-Z0-9]+)\\.(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)\\+[-0-9a-z]*)?$"
+export KUBE_CI_VERSION_DASHED_REGEX="^v(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-([a-zA-Z0-9]+)-(0|[1-9][0-9]*)(-(0|[1-9][0-9]*)\\+[-0-9a-z]*)?"
```
Comment (BenTheElder) [inline cluster/common.sh]: why are we exporting these?

File: cluster/common.sh (hunk 6 above)
Comment (beautytiger) [reply cluster/common.sh]: Yes, this variable is used in hack/get-build.sh. I think it's better to export it.

File: cluster/common.sh (hunk 7 above)
Comment (beautytiger) [reply cluster/common.sh]: Good lesson!

File: cluster/common.sh (hunk 8 above)
Comment (beautytiger) [reply cluster/common.sh]: KUBE_CI_VERSION_REGEX is used in hack/get-build.sh and KUBE_CI_VERSION_DASHED_REGEX is used in cluster/gce/util.sh

File: cluster/common.sh (hunk 2 above)
Comment (beautytiger) [reply cluster/common.sh]: these 2 variable are referenced in this function, but they are defined outside, we should get their value first:
```bash
  if [[ -n "${GEN_ETCD_CA_CERT}" && -n "${GEN_ETCD_CA_KEY}" ]]; then
    echo "${ca_cert}" | base64 --decode | gunzip > ca.pem
    echo "${ca_key}" | base64 --decode > ca-key.pem
  fi
```

File: cluster/common.sh (hunk 1 above)
Comment (beautytiger) [reply cluster/common.sh]: Good point, so I change to just suppress this warning, it seems ok. What do you think?

File: cluster/common.sh (hunk 3 above)
Comment (beautytiger) [reply cluster/common.sh]: This function's return value is used in if statements as below:
```bash
function verify-release-tars() {
  if ! find-release-tars; then
    download-release-binaries
  fi
}
```
I can not just export these variable in the next line, or the function will always return 0.
Do you have better solution?

File: cluster/common.sh (hunk 4 above)
Comment (beautytiger) [reply cluster/common.sh]: Same as above.

File: cluster/common.sh (hunk 5 above)
Comment (beautytiger) [reply cluster/common.sh]: Same as above.

File: cluster/common.sh [hunk 9]
```diff
@@ -277,7 +282,7 @@ function load-or-gen-kube-basicauth() {
 #   KUBE_VERSION
 function set_binary_version() {
   if [[ "${1}" =~ "/" ]]; then
-    IFS='/' read -a path <<< "${1}"
+    IFS='/' read -a -r path <<< "${1}"
```
Comment (k-toyoda-pi) [inline cluster/common.sh]: `read -a -r` can not work. I think `read -r -a` is correct.

Comment (beautytiger) [reply cluster/common.sh]: Thanks!

File: cluster/common.sh (hunk 6 above)
Comment (BenTheElder) [reply cluster/common.sh]: we should update the comment above to note that it is exported versus just set

File: cluster/common.sh (hunk 4 above)
Comment (BenTheElder) [reply cluster/common.sh]: https://github.com/kubernetes/kubernetes/pull/82454/files#r324942507

File: cluster/common.sh (hunk 5 above)
Comment (BenTheElder) [reply cluster/common.sh]: https://github.com/kubernetes/kubernetes/pull/82454/files#r324942507

File: cluster/common.sh (hunk 2 above)
Comment (BenTheElder) [reply cluster/common.sh]: they're not actually local though, and we probably shouldn't allow unset values

Comment (BenTheElder) [reply cluster/common.sh]: we should not redefine these with the same name nor should we allow unset values for this purpose

File: cluster/common.sh (hunk 6 above)
Comment (beautytiger) [reply cluster/common.sh]: Yes and done.

File: cluster/common.sh (hunk 2 above)
Comment (beautytiger) [reply cluster/common.sh]: Keep as it was and update comment, please check.

File: cluster/common.sh (hunk 3 above)
Comment (beautytiger) [reply cluster/common.sh]: update comment

File: cluster/common.sh (hunk 4 above)
Comment (beautytiger) [reply cluster/common.sh]: update in comment

File: cluster/common.sh (hunk 5 above)
Comment (beautytiger) [reply cluster/common.sh]: update in comment

File: cluster/common.sh (hunk 8 above)
Comment (BenTheElder) [reply cluster/common.sh]: ok, let's mention that 👍

File: cluster/common.sh (hunk 3 above)
Comment (BenTheElder) [reply cluster/common.sh]: huh, if you export on the next line it will fail on this line if it fails and not get to the export line?

File: cluster/common.sh [hunk 10]
```diff
@@ -423,7 +435,9 @@ EOF
   fi
 
   if [[ -n "${GEN_ETCD_CA_CERT}" && -n "${GEN_ETCD_CA_KEY}" ]]; then
+    # shellcheck disable=SC2154
```
Comment (BenTheElder) [inline cluster/common.sh]: please don't disable without leaving a comment explaining why we chose to disable a check, so we may fix it in the future or understand that it is actually correct (rare).

Comment (beautytiger) [reply cluster/common.sh]: good point!

File: cluster/common.sh (hunk 3 above)
Comment (beautytiger) [reply cluster/common.sh]: I modified this function `find-release-tars`, split the vars and it return value, it is a little verbose, please check again.

File: cluster/common.sh (hunk 8 above)
Comment (beautytiger) [reply cluster/common.sh]: update in comment

File: cluster/common.sh [hunk 11]
```diff
@@ -29,15 +29,18 @@ source "${KUBE_ROOT}/hack/lib/util.sh"
 #
 # NOTE This must match the version_regex in build/common.sh
 # kube::release::parse_and_validate_release_version()
-KUBE_RELEASE_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(-([a-zA-Z0-9]+)\\.(0|[1-9][0-9]*))?$"
-KUBE_RELEASE_VERSION_DASHED_REGEX="v(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-(0|[1-9][0-9]*)(-([a-zA-Z0-9]+)-(0|[1-9][0-9]*))?"
+export KUBE_RELEASE_VERSION_REGEX="^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(-([a-zA-Z0-9]+)\\.(0|[1-9][0-9]*))?$"
+export KUBE_RELEASE_VERSION_DASHED_REGEX="v(0|[1-9][0-9]*)-(0|[1-9][0-9]*)-(0|[1-9][0-9]*)(-([a-zA-Z0-9]+)-(0|[1-9][0-9]*))?"
```
Comment (BenTheElder) [inline cluster/common.sh]: and these?

File: cluster/common.sh (hunk 3 above)
Comment (BenTheElder) [reply cluster/common.sh]: still disabling?

File: cluster/common.sh [hunk 12]
```diff
@@ -475,7 +518,7 @@ function verify-kube-binaries() {
 # If KUBERNETES_SKIP_CONFIRM is set to y, we'll automatically download binaries
 # without prompting.
 function verify-release-tars() {
-  if ! $(find-release-tars); then
+  if ! find-release-tars; then
```
Comment (BenTheElder) [inline cluster/common.sh]: this is a behavioral change, is find-release-tars in the current shell ok? the previous version runs in a subshell which means EG the environment variables are isolated.

Comment (beautytiger) [reply cluster/common.sh]: I examine the project, there is no variable overlapping in this function, this change should be ok.

Comment (BenTheElder) [reply cluster/common.sh]: what about `exit` ?

Comment (BenTheElder) [reply cluster/common.sh]: looks like that's not an issue
